container/client.go

```diff
@@ -0,0 +1,202 @@
+	minor  string
+}
+
+// NewDockerClient returns dockerClient instance ready for communication with the server endpoint `unix:///var/run/docker.sock`
+func NewDockerClient(endpoint string) (*DockerClient, error) {
+	client, err := docker.NewClient(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot initialize docker client instance with the given endpoint `%s`, err=%v", endpoint, err)
+	}
+
+	err = client.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	dc := &DockerClient{
+		cl:           client,
+		inspectCache: map[string]*docker.Container{},
+		//TODO: handle fs stats
+		//diskUsageCollector: fs.DiskUsageCollector{},
+	}
+
+	// TODO: handle fs stats
+	//dc.diskUsageCollector.Init()
+
+	return dc, nil
+}
+
+// InspectContainer returns details information about running container
+func (dc *DockerClient) InspectContainer(id string) (*docker.Container, error) {
+	dc.inspectMutex.Lock()
+	defer dc.inspectMutex.Unlock()
+
+	// check if the inspect info is already stored in inspectCache
+	if info, haveInfo := dc.inspectCache[id]; haveInfo {
+		return info, nil
+	}
+
+	info, err := dc.cl.InspectContainer(id)
+	if err != nil {
+		return nil, err
+	}
+	dc.inspectCache[id] = info
+
+	return info, nil
+}
+
+// GetDockerParam returns given parameter value fro running docker engine
+func (dc *DockerClient) GetDockerParam(param string) (string, error) {
+	env, err := dc.cl.Info()
+	if err != nil {
+		return "", err
+	}
+
+	if env.Exists(param) {
+		return env.Get(param), nil
+	}
+
+	return "", fmt.Errorf("%s not found", param)
+}
+
+// GetShortID returns short container ID (12 chars)
+func GetShortID(dockerID string) (string, error) {
+	if dockerID == "root" {
+		return dockerID, nil
+	}
+
+	if len(dockerID) < 12 {
+		return "", fmt.Errorf("Docker id %v is too short (the length of id should equal at least 12)", dockerID)
+	}
+
+	return dockerID[:12], nil
+}
+
+// ListContainersAsMap returns list of all available docker containers and base information about them (status, uptime, etc.)
+func (dc *DockerClient) ListContainersAsMap() (map[string]*ContainerData, error) {
+	containers := make(map[string]*ContainerData)
+
+	containerList, err := dc.cl.ListContainers(docker.ListContainersOptions{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range containerList {
+		shortID, err := GetShortID(c.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		spec := Specification{
+			Status:     c.Status,
+			Created:    time.Unix(c.Created, 0).Format("2006-01-02T15:04:05Z07:00"),
+			Image:      c.Image,
+			SizeRw:     c.SizeRw,
+			SizeRootFs: c.SizeRootFs,
+			Labels:     c.Labels,
+		}
+
+		containerData := ContainerData{
+			ID:            c.ID,
+			Specification: spec,
+			Stats:         NewStatistics(),
+		}
+
+		containers[shortID] = &containerData
+	}
+
+	containers["root"] = &ContainerData{ID: "/", Stats: NewStatistics()}
+
+	if len(containers) == 0 {
+		return nil, fmt.Errorf("No docker containers found %d", len(containers))
+	}
```

Comment:
I noticed that this `if` statement does not have sense because this `containers` map always contains "root". So, what do you think about moving this statement before adding "root" to this map, but only logging this fact as an "INFO" or "DEBUG", because even there is no docker containers, collection should not fail as it works currently (in a sense by accident).